### DISCUSSION TITLE
Error regarding NO_EXT_SEQ

### DISCUSSION THREAD

**Discussion by misch42** [strongswan/strongswan#1399] [General]
Hi,

the doc for swanctl.conf says:

proposals = this includes IKE an encryption algorithm, an integrity algorithm, a pseudo random function (PRF) and a Diffie-Hellman key exchange group.

esp_proposals = this includes an integrity algorithm, an encryption algorithm, an optional Diffie-Hellman group and an optional Extended Sequence Number Mode (ESN) indicator.

Isn't is a bit disturbing that in the first option we have encryption first and then integrity and in the esp option we have first the integrity and then the encryption? Confuses me every time ...

Is the a list of all keywords that can be used? I know of the Cipher suits page of IKEv2, but there is no list of the ESN mode keywords, or at least I did not find it.

My problem: I have a working ipsec.conf based config that agrees on a selected proposal:

ESP:AES_CBC_256/HMAC_SHA2_256_128/ECP_384/NO_EXT_SEQ

How can I tell swanctl.conf to use the "NO_EXT_SEQ" ESN mode?

a simple esp_proposals = sha256-aes256-ecp384 will not work. Strongswan sends some PRF function that the other side cannot agree upon. Any idea?

I found some discussions that indicate an error in creating the proposal triplet / quadruplet. Might this be the case?

Michael Schwartzkopff

**Comment by misch42** [comment]
Ok,

basically I have the problem that the child is not established:

Config:
esp_proposals = sha256-aes256-ecp384

Logs:
14[CFG] vici initiate CHILD_SA 'my-net'
08[IKE] initiating IKE_SA my[24] to x.x.x.x
08[ENC] generating IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) ]
08[NET] sending packet: from y.y.y.y.254[500] to x.x.x.x[500] (304 bytes)
16[NET] received packet: from x.x.x.x[500] to y.y.y.y[500] (423 bytes)
16[ENC] parsed IKE_SA_INIT response 0 [ SA KE No V V CERTREQ N(FRAG_SUP) V ]
16[IKE] received Cisco Delete Reason vendor ID
16[IKE] received Cisco Copyright (c) 2009 vendor ID
16[IKE] received FRAGMENTATION vendor ID
16[CFG] selected proposal: IKE:AES_CBC_256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/ECP_384
16[IKE] received 3 cert requests for an unknown ca
16[IKE] authentication of 'y.y.y.y' (myself) with pre-shared key
16[IKE] establishing CHILD_SA my-net{25}
16[ENC] generating IKE_AUTH request 1 [ IDi N(INIT_CONTACT) IDr AUTH SA TSi TSr N(MOBIKE_SUP) N(ADD_4_ADDR) N(EAP_ONLY) N(MSG_ID_SYN_SUP) ]
16[NET] sending packet: from y.y.y.y[500] to x.x.x.x[500] (272 bytes)
06[NET] received packet: from x.x.x.x[500] to y.y.y.y[500] (160 bytes)
06[ENC] parsed IKE_AUTH response 1 [ V IDr AUTH N(NO_PROP) ]
06[IKE] authentication of 'x.x.x.x' with pre-shared key successful
06[IKE] IKE_SA my[24] established between y.y.y.y[y.y.y.y]...x.x.x.x[x.x.x.x]
06[IKE] scheduling reauthentication in 26949s
06[IKE] maximum IKE_SA lifetime 29829s
06[IKE] received NO_PROPOSAL_CHOSEN notify, no CHILD_SA built
0


with a ipsec.conf based config the logs read:
charon-custom[223514]: 07[ENC] generating CREATE_CHILD_SA request 2 [ SA No KE TSi TSr ]
charon-custom[223514]: 07[NET] sending packet: from y.y.y.y[500] to x.x.x.x[500] (464 bytes)
ipsec[223514]: 08[NET] received packet: from x.x.x.x[500] to y.y.y.y[500] (352 bytes)
ipsec[223514]: 08[ENC] parsed CREATE_CHILD_SA response 2 [ SA No KE TSi TSr ]
charon-custom[223514]: 08[NET] received packet: from x.x.x.x500] to y.y.y.y[500] (352 bytes)
ipsec[223514]: 08[CFG] selected proposal: ESP:AES_CBC_256/HMAC_SHA2_256_128/ECP_384/NO_EXT_SEQ
ipsec[223514]: 08[IKE] CHILD_SA sa_0{2} established with SPIs c547f0c1_i 56705678_o and TS ....

Ideas?

**Comment by Thermi** [comment]
> Isn't is a bit disturbing that in the first option we have encryption first and then integrity and in the esp option we have first the integrity and then the encryption? Confuses me every time ...

You don't need to do it. Just use enc-int-dh.

> Is the a list of all keywords that can be used? I know of the Cipher suits page of IKEv2, but there is no list of the ESN mode keywords, or at least I did not find it.

It's just -esn or -noesn.

From the man page for swanctl.conf:
```
       connections.<conn>.children.<child>.esp_proposals [default]
[...]
              Extended Sequence Number support may be indicated with the esn and noesn values, both may be included to indicate support
              for both modes. If omitted, noesn is assumed.
```

a simple esp_proposals = sha256-aes256-ecp384 will not work. Strongswan sends some PRF function that the other side cannot agree upon. Any idea?

ESP does not use a PRF, that is IKE only. Just use aes256-sha256-ecp384 as esp_proposal, like the one for ipsec.conf indicates.
I suppose the CISCO does not handle the order of int-enc-dh correctly.

What is your exact swanctl.conf and ipsec.conf? Maybe the key exchange version is something they do not like.

**Reply by tobiasbrunner** [reply]
> > Isn't is a bit disturbing that in the first option we have encryption first and then integrity and in the esp option we have first the integrity and then the encryption? Confuses me every time ...
> 
> You don't need to do it. Just use enc-int-dh.

Yeah, the transforms types don't have to be ordered in any specific way (i.e. you could write `enc1-int1-dh1-int2-int3-enc2-dh2-dh3-enc3` if you like, which is the same as `enc1-enc2-enc3-int1-int2-int3-dh1-dh2-dh3`), the order of algorithms only matters in regards to preference. 

Also, the syntax is basically the same for ipsec.conf and swanctl.conf. Except for the behavior regarding default proposal, which ipsec.conf added if `! ` was not used, but has to be added explicitly with `default` in swanctl.conf.

> I suppose the CISCO does not handle the order of int-enc-dh correctly.

On the wire, the transforms are always encoded in the same order (ENCR, INTEG, PRF, KE, ESN, I don't think the RFC demands any specific order, it's just how strongSwan does it). So the order when configuring really only matters per transform type as that can determine the selected algorithm. Those listed first are preferred, however, it's not certain the responder selects the first one even if it supports it, as it may prefer its own configured proposal and algorithm order, which strongSwan does by default.
